File: backend/src/storage/configStore.ts
/**
 * 配置存储模块
 * 使用 Redis 存储网关配置
 */

import { getRedisClient } from './redis.js'
import { createLogger } from '../utils/logger.js'

const logger = createLogger('ConfigStore')

// Redis Key
const CONFIG_KEY = 'config'
const SELECTED_ACCOUNTS_KEY = 'config:selectedAccounts'
const API_KEYS_KEY = 'config:apiKeys'

export interface GatewayConfig {
  // 服务配置
  port: number
  host: string

  // 代理配置
  enableMultiAccount: boolean
  maxConcurrent: number
  maxRetries: number
  retryDelay: number
  requestTimeout: number
  preferredEndpoint?: 'codewhisperer' | 'amazonq'

  // 限流配置
  rateLimitEnabled: boolean
  rateLimitWindow: number
  rateLimitMax: number
}

const DEFAULT_CONFIG: GatewayConfig = {
  port: 3000,
  host: '0.0.0.0',
  enableMultiAccount: true,
  maxConcurrent: 5,
  maxRetries: 3,
  retryDelay: 1000,
  requestTimeout: 120000,
  rateLimitEnabled: false,
  rateLimitWindow: 60000,
  rateLimitMax: 100
}

/**
 * 获取网关配置
 */
export async function getGatewayConfig(): Promise<GatewayConfig> {
  const redis = getRedisClient()

  try {
    const data = await redis.hgetall(CONFIG_KEY)

    if (!data || Object.keys(data).length === 0) {
      return { ...DEFAULT_CONFIG }
    }

    return {
      port: parseInt(data.port, 10) || DEFAULT_CONFIG.port,
      host: data.host || DEFAULT_CONFIG.host,
      enableMultiAccount: data.enableMultiAccount === 'true',
      maxConcurrent: parseInt(data.maxConcurrent, 10) || DEFAULT_CONFIG.maxConcurrent,
      maxRetries: parseInt(data.maxRetries, 10) || DEFAULT_CONFIG.maxRetries,
      retryDelay: parseInt(data.retryDelay, 10) || DEFAULT_CONFIG.retryDelay,
      requestTimeout: parseInt(data.requestTimeout, 10) || DEFAULT_CONFIG.requestTimeout,
      preferredEndpoint: data.preferredEndpoint as GatewayConfig['preferredEndpoint'],
      rateLimitEnabled: data.rateLimitEnabled === 'true',
      rateLimitWindow: parseInt(data.rateLimitWindow, 10) || DEFAULT_CONFIG.rateLimitWindow,
      rateLimitMax: parseInt(data.rateLimitMax, 10) || DEFAULT_CONFIG.rateLimitMax
    }
  } catch (error) {
    logger.error('Failed to get gateway config', { error: (error as Error).message })
    return { ...DEFAULT_CONFIG }
  }
}

/**
 * 更新网关配置
 */
export async function updateGatewayConfig(updates: Partial<GatewayConfig>): Promise<GatewayConfig> {
  const redis = getRedisClient()

  try {
    const current = await getGatewayConfig()
    const updated = { ...current, ...updates }

    const data: Record<string, string> = {}
    for (const [key, value] of Object.entries(updated)) {
      if (value !== undefined && value !== null) {
        data[key] = String(value)
      }
    }

    await redis.hset(CONFIG_KEY, data)
    logger.info('Gateway config updated', { updates })

    return updated
  } catch (error) {
    logger.error('Failed to update gateway config', { error: (error as Error).message })
    throw error
  }
}

/**
 * 获取选中的账号 ID 列表
 */
export async function getSelectedAccounts(): Promise<string[]> {
  const redis = getRedisClient()

  try {
    return await redis.smembers(SELECTED_ACCOUNTS_KEY)
  } catch (error) {
    logger.error('Failed to get selected accounts', { error: (error as Error).message })
    return []
  }
}

/**
 * 设置选中的账号
 */
export async function setSelectedAccounts(accountIds: string[]): Promise<void> {
  const redis = getRedisClient()

  try {
    await redis.del(SELECTED_ACCOUNTS_KEY)
    if (accountIds.length > 0) {
      await redis.sadd(SELECTED_ACCOUNTS_KEY, ...accountIds)
    }
    logger.info('Selected accounts updated', { count: accountIds.length })
  } catch (error) {
    logger.error('Failed to set selected accounts', { error: (error as Error).message })
    throw error
  }
}

/**
 * 添加选中账号
 */
export async function addSelectedAccount(accountId: string): Promise<void> {
  const redis = getRedisClient()

  try {
    await redis.sadd(SELECTED_ACCOUNTS_KEY, accountId)
  } catch (error) {
    logger.error('Failed to add selected account', { error: (error as Error).message })
    throw error
  }
}

/**
 * 移除选中账号
 */
export async function removeSelectedAccount(accountId: string): Promise<void> {
  const redis = getRedisClient()

  try {
    await redis.srem(SELECTED_ACCOUNTS_KEY, accountId)
  } catch (error) {
    logger.error('Failed to remove selected account', { error: (error as Error).message })
    throw error
  }
}

// ============ API Key 管理 ============

export interface ApiKeyRecord {
  id: string
  key: string
  name: string
  createdAt: number
  lastUsed?: number
}

/**
 * 获取所有 API Key
 */
export async function getAllApiKeys(): Promise<ApiKeyRecord[]> {
  const redis = getRedisClient()

  try {
    const data = await redis.hgetall(API_KEYS_KEY)

    if (!data || Object.keys(data).length === 0) {
      return []
    }

    return Object.values(data).map(json => JSON.parse(json) as ApiKeyRecord)
  } catch (error) {
    logger.error('Failed to get API keys', { error: (error as Error).message })
    return []
  }
}

/**
 * 添加 API Key
 */
export async function addApiKey(record: ApiKeyRecord): Promise<void> {
  const redis = getRedisClient()

  try {
    await redis.hset(API_KEYS_KEY, record.id, JSON.stringify(record))
    logger.info('API key added', { id: record.id, name: record.name })
  } catch (error) {
    logger.error('Failed to add API key', { error: (error as Error).message })
    throw error
  }
}

/**
 * 删除 API Key
 */
export async function deleteApiKey(id: string): Promise<boolean> {
  const redis = getRedisClient()

  try {
    const result = await redis.hdel(API_KEYS_KEY, id)
    if (result > 0) {
      logger.info('API key deleted', { id })
      return true
    }
    return false
  } catch (error) {
    logger.error('Failed to delete API key', { error: (error as Error).message })
    return false
  }
}

/**
 * 验证 API Key
 */
export async function validateApiKey(key: string): Promise<boolean> {
  const redis = getRedisClient()

  try {
    const data = await redis.hgetall(API_KEYS_KEY)

    if (!data || Object.keys(data).length === 0) {
      // 没有配置任何 API Key，允许访问
      return true
    }

    for (const json of Object.values(data)) {
      const record = JSON.parse(json) as ApiKeyRecord
      if (record.key === key) {
        // 更新最后使用时间
        record.lastUsed = Date.now()
        await redis.hset(API_KEYS_KEY, record.id, JSON.stringify(record))
        return true
      }
    }

    return false
  } catch (error) {
    logger.error('Failed to validate API key', { error: (error as Error).message })
    return false
  }
}

File: packages/gateway/storage/config_store.go
package storage

import (
	"context"
	"fmt"
	"strconv"
	"strings"

	"github.com/rs/zerolog/log"

	"github.com/kiro-gateway/gateway/model"
)

// Redis key constants for configuration storage.
const (
	configKey           = "config"
	selectedAccountsKey = "config:selectedAccounts"
	apiKeysKey          = "config:apiKeys"
	apiKeysIndexKey     = "config:apiKeyIndex"
	webhookConfigKey    = "webhook_config"
)

func safeParseInt(s string, fallback int) int {
	if s == "" {
		return fallback
	}
	v, err := strconv.Atoi(s)
	if err != nil {
		return fallback
	}
	return v
}

func safeParseFloat(s string, fallback float64) float64 {
	if s == "" {
		return fallback
	}
	v, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return fallback
	}
	return v
}

func safeParseBoolStr(s string, fallback bool) bool {
	if s == "" {
		return fallback
	}
	return s == "true"
}

func intPtr(v int) *int           { return &v }
func boolPtr(v bool) *bool        { return &v }
func strPtr(s string) *string     { return &s }
func floatPtr(v float64) *float64 { return &v }

func coalesce(values ...string) string {
	for _, v := range values {
		if v != "" {
			return v
		}
	}
	return ""
}

func coalesceStr(values ...string) string {
	for _, v := range values {
		if strings.TrimSpace(v) != "" {
			return v
		}
	}
	return ""
}

// defaultGatewayConfig returns a GatewayConfig with TS backend defaults.
func defaultGatewayConfig() *model.GatewayConfig {
	return &model.GatewayConfig{
		Port:                       3000,
		Host:                       "0.0.0.0",
		EnableMultiAccount:         true,
		MaxConcurrent:              0,
		MaxRetries:                 3,
		RetryDelay:                 1000,
		RequestTimeout:             120000,
		DefaultRegion:              strPtr("us-east-1"),
		DisableTools:               boolPtr(false),
		AutoContinueRounds:         intPtr(0),
		AutoSwitchOnQuotaExhausted: boolPtr(true),
		RateLimitEnabled:           false,
		RateLimitWindow:            60000,
		RateLimitMax:               100,
		TestModelID:                strPtr("claude-sonnet-4.5"),
	}
}

// GetGatewayConfig retrieves the full gateway configuration from Redis.
// Handles field aliases for backward compatibility with the TS backend.
func GetGatewayConfig(ctx context.Context) (*model.GatewayConfig, error) {
	rdb := GetClient()
	def := defaultGatewayConfig()

	data, err := rdb.HGetAll(ctx, Key(configKey)).Result()
	if err != nil {
		return def, fmt.Errorf("get gateway config: %w", err)
	}
	if len(data) == 0 {
		return def, nil
	}

	// Field alias handling: prefer standard names, fall back to aliases.
	enableMultiRaw := coalesce(data["enableMultiAccount"], data["multiAccountEnabled"])
	enableLoggingRaw := coalesce(data["enableRequestLogging"], data["logRequests"])
	tokenRefreshRaw := coalesce(data["tokenRefreshBeforeExpiry"], data["tokenRefreshAdvance"])
	autoContinueRaw := coalesce(data["autoContinueRounds"], data["toolCallAutoRounds"])
	disableToolsVal := data["disableTools"] == "true" || data["disableToolCalls"] == "true"
	autoContinueVal := safeParseInt(autoContinueRaw, 0)

	cfg := &model.GatewayConfig{
		Port:               safeParseInt(data["port"], def.Port),
		Host:               coalesceStr(data["host"], def.Host),
		EnableMultiAccount: safeParseBoolStr(enableMultiRaw, def.EnableMultiAccount),
		MaxConcurrent:      safeParseInt(data["maxConcurrent"], def.MaxConcurrent),
		MaxRetries:         safeParseInt(data["maxRetries"], def.MaxRetries),
		RetryDelay:         safeParseInt(data["retryDelay"], def.RetryDelay),
		RequestTimeout:     safeParseInt(data["requestTimeout"], def.RequestTimeout),
		RateLimitEnabled:   data["rateLimitEnabled"] == "true",
		RateLimitWindow:    safeParseInt(data["rateLimitWindow"], def.RateLimitWindow),
		RateLimitMax:       safeParseInt(data["rateLimitMax"], def.RateLimitMax),
	}

	// Pointer-based optional fields.
	if v := data["preferredEndpoint"]; v != "" {
		cfg.PreferredEndpoint = strPtr(v)
	}
	cfg.DefaultRegion = strPtr(coalesceStr(data["defaultRegion"], "us-east-1"))
	cfg.DisableTools = boolPtr(disableToolsVal)
	cfg.DisableToolCalls = boolPtr(disableToolsVal)
	cfg.AutoContinueRounds = intPtr(autoContinueVal)
	cfg.ToolCallAutoRounds = intPtr(autoContinueVal)

	autoSwitch := true
	if v, ok := data["autoSwitchOnQuotaExhausted"]; ok {
		autoSwitch = v == "true"
	}
	cfg.AutoSwitchOnQuotaExhausted = boolPtr(autoSwitch)

	loggingEnabled := true
	if enableLoggingRaw != "" {
		loggingEnabled = enableLoggingRaw == "true"
	}
	cfg.EnableRequestLogging = boolPtr(loggingEnabled)
	cfg.TokenRefreshBeforeExpiry = intPtr(safeParseInt(tokenRefreshRaw, 300))

	if v := data["queueEnabled"]; v != "" {
		cfg.QueueEnabled = boolPtr(v == "true")
	}
	if v := data["testModelId"]; v != "" {
		cfg.TestModelID = strPtr(v)
	} else {
		cfg.TestModelID = strPtr("claude-sonnet-4.5")
	}

	// Optional account pool settings.
	if v := data["errorCooldownTime"]; v != "" {
		cfg.ErrorCooldownTime = intPtr(safeParseInt(v, 0))
	}
	if v := data["maxConsecutiveErrors"]; v != "" {
		cfg.MaxConsecutiveErrors = intPtr(safeParseInt(v, 0))
	}
	if v := data["quotaResetTime"]; v != "" {
		cfg.QuotaResetTime = intPtr(safeParseInt(v, 0))
	}
	if v := data["autoStopErrorCodes"]; v != "" {
		cfg.AutoStopErrorCodes = strPtr(v)
	}
	if v := data["autoStopErrorPatterns"]; v != "" {
		cfg.AutoStopErrorPatterns = strPtr(v)
	}
	if v := data["quotaUsageThreshold"]; v != "" {
		cfg.QuotaUsageThreshold = intPtr(safeParseInt(v, 0))
	}

	// Queue settings.
	if v := data["queueMaxSize"]; v != "" {
		cfg.QueueMaxSize = intPtr(safeParseInt(v, 0))
	}
	if v := data["queueTimeoutMs"]; v != "" {
		cfg.QueueTimeoutMs = intPtr(safeParseInt(v, 0))
	}

	// Dynamic concurrency.
	if v := data["concurrencyMultiplier"]; v != "" {
		cfg.ConcurrencyMultiplier = floatPtr(safeParseFloat(v, 0))
	}
	if v := data["queueSizeMultiplier"]; v != "" {
		cfg.QueueSizeMultiplier = floatPtr(safeParseFloat(v, 0))
	}

	// Account wait queue.
	if v, ok := data["accountWaitEnabled"]; ok {
		cfg.AccountWaitEnabled = boolPtr(v == "true")
	}
	if v := data["accountWaitTimeoutMs"]; v != "" {
		cfg.AccountWaitTimeoutMs = intPtr(safeParseInt(v, 0))
	}
	if v := data["accountWaitMaxSize"]; v != "" {
		cfg.AccountWaitMaxSize = intPtr(safeParseInt(v, 0))
	}

	// Session stickiness.
	if v, ok := data["sessionEnabled"]; ok {
		cfg.SessionEnabled = boolPtr(v == "true")
	}
	if v := data["sessionTtlSeconds"]; v != "" {
		cfg.SessionTTLSeconds = intPtr(safeParseInt(v, 0))
	}
	if v := data["sessionRenewalThresholdSeconds"]; v != "" {
		cfg.SessionRenewalThresholdSeconds = intPtr(safeParseInt(v, 0))
	}
	if v, ok := data["sessionEnableCacheControl"]; ok {
		cfg.SessionEnableCacheControl = boolPtr(v == "true")
	}

	// Proxy service settings.
	if v, ok := data["proxyEnabled"]; ok {
		cfg.ProxyEnabled = boolPtr(v == "true")
	}
	if v := data["proxyPort"]; v != "" {
		cfg.ProxyPort = intPtr(safeParseInt(v, 0))
	}
	if v := data["proxyHost"]; v != "" {
		cfg.ProxyHost = strPtr(v)
	}

	// Automation.
	if v, ok := data["autoStart"]; ok {
		cfg.AutoStart = boolPtr(v == "true")
	}

	return cfg, nil
}

// UpdateGatewayConfig merges updates into the stored configuration.
// Normalizes field aliases and cleans up stale alias keys.
func UpdateGatewayConfig(ctx context.Context, updates map[string]interface{}) (*model.GatewayConfig, error) {
	rdb := GetClient()

	// Normalize field aliases to canonical names.
	if v, ok := updates["multiAccountEnabled"]; ok {
		updates["enableMultiAccount"] = v
		delete(updates, "multiAccountEnabled")
	}
	if v, ok := updates["logRequests"]; ok {
		updates["enableRequestLogging"] = v
		delete(updates, "logRequests")
	}
	if v, ok := updates["tokenRefreshAdvance"]; ok {
		updates["tokenRefreshBeforeExpiry"] = v
		delete(updates, "tokenRefreshAdvance")
	}

	// Build flat string map for HSET.
	fields := make(map[string]interface{}, len(updates))
	for k, v := range updates {
		if v == nil {
			continue
		}
		switch val := v.(type) {
		case string:
			fields[k] = val
		case bool:
			if val {
				fields[k] = "true"
			} else {
				fields[k] = "false"
			}
		case int:
			fields[k] = strconv.Itoa(val)
		case int64:
			fields[k] = strconv.FormatInt(val, 10)
		case float64:
			fields[k] = strconv.FormatFloat(val, 'f', -1, 64)
		default:
			fields[k] = fmt.Sprintf("%v", val)
		}
	}

	// Remove stale alias keys to avoid duplicate semantics.
	aliasKeys := []string{"multiAccountEnabled", "logRequests", "tokenRefreshAdvance"}
	for _, ak := range aliasKeys {
		rdb.HDel(ctx, Key(configKey), ak)
	}

	if len(fields) > 0 {
		if err := rdb.HSet(ctx, Key(configKey), fields).Err(); err != nil {
			return nil, fmt.Errorf("update gateway config: %w", err)
		}
	}

	log.Info().Msg("Gateway config updated")
	return GetGatewayConfig(ctx)
}

// GetSelectedAccounts returns the set of selected account IDs.
func GetSelectedAccounts(ctx context.Context) ([]string, error) {
	return GetClient().SMembers(ctx, Key(selectedAccountsKey)).Result()
}

// SetSelectedAccounts replaces the entire set of selected account IDs.
func SetSelectedAccounts(ctx context.Context, ids []string) error {
	rdb := GetClient()

	rdb.Del(ctx, Key(selectedAccountsKey))
	if len(ids) > 0 {
		members := make([]interface{}, len(ids))
		for i, id := range ids {
			members[i] = id
		}
		if err := rdb.SAdd(ctx, Key(selectedAccountsKey), members...).Err(); err != nil {
			return fmt.Errorf("set selected accounts: %w", err)
		}
	}

	log.Info().Int("count", len(ids)).Msg("Selected accounts updated")
	return nil
}